doc/proposals/validator-library.md

```diff
@@ -0,0 +1,201 @@
+
+One closure function around one object is a unit of validation. This discretization allows developers to break up validation logic into a comprehensible manner and turn validators on/off by modifying single lines of code. Library users can compose multiple validator units into one `Validators` object, and run them all at once on one or more objects. Running validators can be done lazily (collect validator closures and run them later) or actively (invoke the closure at the point of its return).
+
+#### Errors
+
+An `errors` package will be provided that exposes custom result types for validation. The `ManifestResult` type will be returned by each closure function, containing sets of `Error`'s and `Warning`'s. The underlying `Error` type for both sets will be similar to [`field.Error`][crd-validation-error]. These results will help end users discover what needs fixing in their manifests.
+
+#### Default `Validator`'s
+
+A default set of `Validator` implementers will be available:
+
+- `ClusterServiceVersion`: validate `ClusterServiceVersion` objects.
+- `CustomResourceDefinition`: validate `CustomResourceDefinition` objects.
+- Package manifest: validate package manifest objects.
+- Bundle: validate all objects in a bundle directory (via [`*registry.Bundle`][registry-bundle-type]) collectively, verifying relationships between objects.
+- Manifests: validate all objects in a manifests directory collectively, verifying relationships between objects.
+
+Each will be exported as a `Default{Type}Validators` function containing a set of validators disjointed from other default validators. These functions should take a variadic `...Validator` parameter, such that a user can add custom validators to the list of defaults provided.
+
+### Implementation details
+
+`Validator` interface:
+
+```go
+package validator
+
+import (
+	"github.com/operator-framework/api/pkg/validation/errors"
+)
+
+// ValidatorFunc returns a ManifestResult containing errors and warnings from
+// validating some object. These are typically closures or methods.
+type ValidatorFunc func() errors.ManifestResult
+
+// ValidationFuncs is a set of validation functions.
+type ValidatorFuncs []ValidatorFunc
+
+// Validator is an interface for validating arbitrary objects. A Validator
+// returns a set of functions that each validate some underlying object.
+// This allows the implementer to easily break down each validation step
+// into discrete units, and perform either lazy or active validation.
+type Validator interface {
+	// GetFuncs takes a list of arbitrary objects and returns a set of functions
+	// that each validate some object from the list.
+	GetFuncs(...interface{}) ValidatorFuncs
```

Comment:
Would it be simpler to deal with one object at a time? (or async) such that for a given object we get the functions and run them, may make it easier if the `validateCRD` func conformed to the interface instead of wrapping everything in a closure? 

Would this allow for a simpler validate function code?

Reply:
I agree. I've updated #3 to use @njhale's suggested simplified `Validator` interface. Once that has been agreed upon I'll update this proposal, or close this PR if we find it unnecessary.